match-derive/src/lib.rs

```diff
@@ -1,7 +1,179 @@
-extern crate proc_macro;
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::parse_macro_input;
 
-#[proc_macro_derive(Matcher)]
-pub fn derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
-    let _ = input;
-    unimplemented!()
+mod parse;
+
+use parse::MatcherDerive;
+
+#[proc_macro_derive(Matcher, attributes(matcher))]
+pub fn derive(input: TokenStream) -> TokenStream {
+    // TODO when we generate a name that isn't a valid ident or is a keyword, generate a different
+    // name rather than panicking.
```

Comment:
I think it should cause a compilation error, but not do any name picking magic.

Reply:
There is prior art for this in `wasm-bindgen`, where they append a `_` (underscore, e.g. `match_`) if there is a name clash. There they have to generate equivalent rust names for wepapi interfaces, which is not exactly the same situation as us, and so we shouldn't necessarily do what they do.